tests/stream_states.rs

```diff
@@ -808,3 +808,65 @@ fn send_data_after_headers_eos() {
 fn exceed_max_streams() {
 }
 */
+
+
+#[test]
+fn rst_while_closing() {
+    // Test to reproduce panic in issue #246 --- receipt of a RST_STREAM frame
+    // on a stream in the Half Closed (remote) state with a queued EOS causes
+    // a panic.
+    let _ = ::env_logger::try_init();
+    let (io, srv) = mock::new();
+
+    let srv = srv.assert_client_handshake()
+        .unwrap()
+        .recv_settings()
+        .recv_frame(
+            frames::headers(1)
+                .request("GET", "https://example.com/")
+        )
+        .send_frame(frames::headers(1).response(200))
+        .send_frame(frames::headers(1).eos())
+        // Idling for a moment here is necessary to ensure that the client
+        // enqueues its TRAILERS frame *before* we send the RST_STREAM frame
+        // which causes the panic.
+        .idle_ms(1)
+        // Send the RST_STREAM frame which causes the client to panic.
+        .send_frame(frames::reset(1).cancel())
+        .recv_frame(frames::reset(1).cancel())
```

Comment:
I believe that, ideally, we would avoid sending a RST_STREAM. While the peer should be able to tolerate this, it is not necessary as a `RST_STREAM` in either direction results in the stream transitioning to closed.

Reply:
Yeah, I think I misunderstood a sentence in the spec saying that we should _allow_ this as saying that we _should_ do this. Had I known that earlier, this PR would have been finished much quicker. My bad!

Corrected this behaviour in 861b10934fb7ac305dbbef020fce9ea3e8c2dda1.